rust/agents/validator/src/submit.rs

```diff
@@ -96,76 +155,85 @@ impl ValidatorSubmitter {
-                        if existing.is_some() {
-                            debug!(
-                                index = queued_checkpoint.index,
-                                "Checkpoint already submitted"
-                            );
-                            continue;
-                        }
-
-                        let signed_checkpoint = self.signer.sign(queued_checkpoint).await?;
-                        self.checkpoint_syncer
-                            .write_checkpoint(&signed_checkpoint)
-                            .await?;
-                        debug!(
-                            index = queued_checkpoint.index,
-                            "Signed and submitted checkpoint"
-                        );
-
-                        // small sleep before signing next checkpoint to avoid rate limiting
-                        sleep(Duration::from_millis(100)).await;
-                    }
-
-                    info!(index = checkpoint.index, "Signed all queued checkpoints");
-
-                    self.metrics
-                        .latest_checkpoint_processed
-                        .set(checkpoint.index as i64);
-
-                    // break out of submitter loop if target checkpoint is reached
-                    reached_target = target_checkpoint.is_some();
-                    break;
-                }
+        debug!(index = checkpoint.index, "Reached tree consistency");
+
+        self.sign_and_submit_checkpoints(checkpoint_queue).await?;
+
+        info!(
+            index = checkpoint.index,
+            "Signed all queued checkpoints until index"
+        );
+
+        Ok(())
+    }
+
+    /// Signs and submits any previously unsubmitted checkpoints.
+    async fn sign_and_submit_checkpoints(
+        &self,
+        checkpoints: Vec<CheckpointWithMessageId>,
+    ) -> Result<()> {
+        for queued_checkpoint in checkpoints {
+            let existing = self
+                .checkpoint_syncer
+                .fetch_checkpoint(queued_checkpoint.index)
+                .await?;
+            if existing.is_some() {
+                debug!(
+                    index = queued_checkpoint.index,
+                    "Checkpoint already submitted"
+                );
+                continue;
             }
 
-            sleep(self.interval).await;
-        }
+            let signed_checkpoint = self.signer.sign(queued_checkpoint).await?;
+            self.checkpoint_syncer
+                .write_checkpoint(&signed_checkpoint)
+                .await?;
+            debug!(
+                index = queued_checkpoint.index,
+                "Signed and submitted checkpoint"
+            );
 
-        // TODO: remove this once validator is tolerant of tasks exiting
-        loop {
-            sleep(Duration::from_secs(u64::MAX)).await;
+            // small sleep before signing next checkpoint to avoid rate limiting
+            sleep(Duration::from_millis(100)).await;
```

Comment:
did you encounter this? assume this is specific to KMS so might be nice not to have it here but not critical

Reply:
No I just copied this from the existing logic, figured this was from that weird KMS thing too so didn't want to change